src/NexusMods.Library/LibraryService.cs

```diff
@@ -0,0 +1,199 @@
+using System.Collections.ObjectModel;
+using System.Reactive.Disposables;
+using System.Reactive.Linq;
+using DynamicData;
+using DynamicData.Kernel;
+using Microsoft.Extensions.Logging;
+using NexusMods.Abstractions.Downloads;
+using NexusMods.Abstractions.FileExtractor;
+using NexusMods.Abstractions.IO.StreamFactories;
+using NexusMods.Abstractions.Library;
+using NexusMods.Hashing.xxHash64;
+using NexusMods.MnemonicDB.Abstractions;
+using NexusMods.Paths;
+
+namespace NexusMods.Library;
+
+/// <summary>
+/// Implementation of <see cref="ILibraryService"/>.
+/// </summary>
+public sealed class LibraryService : ILibraryService, IDisposable
+{
+    private readonly ILogger _logger;
+    private readonly CompositeDisposable _compositeDisposable = new();
+
+    private readonly IConnection _connection;
+    private readonly IFileExtractor _fileExtractor;
+    private readonly TemporaryFileManager _temporaryFileManager;
+
+    private readonly SourceCache<IDownloadActivity, PersistedDownloadStateId> _downloadActivitySourceCache = new(x => x.PersistedStateId);
+    private readonly ReadOnlyObservableCollection<IDownloadActivity> _downloadActivities;
+    public ReadOnlyObservableCollection<IDownloadActivity> DownloadActivities => _downloadActivities;
+
+    /// <summary>
+    /// Constructor.
+    /// </summary>
+    public LibraryService(
+        ILogger<LibraryService> logger,
+        IConnection connection,
+        IFileExtractor fileExtractor,
+        TemporaryFileManager temporaryFileManager)
+    {
+        _logger = logger;
+        _connection = connection;
+        _fileExtractor = fileExtractor;
+        _temporaryFileManager = temporaryFileManager;
+
+        _downloadActivitySourceCache
+            .Connect()
+            .Bind(out _downloadActivities)
+            .Subscribe()
+            .DisposeWith(_compositeDisposable);
+
+        _downloadActivitySourceCache
+            .Connect()
+            .WhenPropertyChanged(downloadActivity => downloadActivity.Status)
+            .Where(propertyValue => propertyValue.Value == PersistedDownloadStatus.Completed)
+            .Select(propertyValue => propertyValue.Sender)
+            .SelectMany(AddCompletedDownloadAsync)
+            .Subscribe()
+            .DisposeWith(_compositeDisposable);
```

Comment:
I don't think there is anything here to handle cancelled activities like we do for Completed ones.
What is the plan for Cancelled downloads? I'm guessing they should no longer have an activity associated with them either.

Reply:
Still trying to figure out what we'd want to use this for. It'll be clearer once we start using this.